src/tree/gpu_hist/histogram.cu

```diff
@@ -145,40 +145,84 @@ __global__ void SharedMemHistKernel(EllpackDeviceAccessor matrix,
   }
 }
 
+template <typename GradientSumT>
+void InitGradientHistogram(int device_idx, int n_bins, HistogramLaunchConfig* config) {
```

Comment:
How expensive is this function? I know some cuda API calls can be very expensive, but I want to do everything possible to avoid storing extra state.

If possible I would call this function before every single kernel invocation and never have to worry about the configuration being passed in an incorrect state.